pkg/cri/server/restart.go

```diff
@@ -152,20 +175,21 @@ func (c *criService) recover(ctx context.Context) error {
 		if status.State == sandboxstore.StateNotReady {
 			continue
 		}
-		controller, err := c.sandboxService.SandboxController(sb.Config, sb.RuntimeHandler)
+		exitCh, err := sb.Sandbox.Wait(ctrdutil.NamespacedContext())
 		if err != nil {
-			log.G(ctx).WithError(err).Error("failed to get sandbox controller while waiting sandbox")
-			continue
-		}
-		exitCh := make(chan containerd.ExitStatus, 1)
-		go func() {
-			exit, err := controller.Wait(ctrdutil.NamespacedContext(), sb.ID)
-			if err != nil {
-				log.G(ctx).WithError(err).Error("failed to wait for sandbox exit")
-				exitCh <- *containerd.NewExitStatus(containerd.UnknownExitStatus, time.Time{}, err)
+			log.G(ctx).WithError(err).Error("failed to wait for sandbox exit")
+			if !errdefs.IsNotFound(err) {
+				return fmt.Errorf("failed to wait for task: %w", err)
 			}
-			exitCh <- *containerd.NewExitStatus(exit.ExitStatus, exit.ExitedAt, nil)
-		}()
+			// Sandbox was in running state, but its task has been deleted,
```

Comment:
if it was unknown before restart... does that mean unknown is not valid after restart.. or are we removing unknown...

I'm probably missing something.

Reply:
I think the status of sandbox is a little bit simpler than container, when we call `ListPodSandbox`, we can only get two states of whether `Ready` or `NotReady`. we only set the sandbox status to `Unknown` in cri cache temporary when we do `RunPodSandbox` and we do not store the state to db. So when containerd restart, it will check the status of sandbox by calling the `Status` or `Wait` APIs of sandbox controller. 
We can also see that in codes of  [recover.go](https://github.com/containerd/containerd/blob/main/pkg/cri/server/podsandbox/recover.go#L94-L114) and [restart.go](https://github.com/containerd/containerd/blob/main/pkg/cri/server/restart.go#L119-L137) that the status of sandbox is set to `Ready` and `NotReady`, without a third state.

Reply:
are you saying this:
https://github.com/containerd/containerd/blob/main/pkg/cri/server/podsandbox/sandbox_run.go#L285-L288](https://github.com/containerd/containerd/blob/main/pkg/cri/store/sandbox/status.go#L76-L78)

only applies to cache.. hmm is that new and we just didn't update the state diagram?  

Ok I missed the move from sandbox having a store that was actually stored to an in memory cache only..

hmm...

Reply:
missed this change: https://github.com/containerd/containerd/pull/7401

Reply:
Should I be concerned here:
https://github.com/containerd/containerd/blob/main/pkg/cri/server/podsandbox/sandbox_run.go#L265-L288
about not using a lock on this cached map (writing directly to the map contents to update state from unknown to ready, and other changes, without doing some sort of lock first via an update api?)

Guessing my confusion is more to the in between status of the cri sandbox store and it's meta information and the moving of that to the new sandbox meta...

Reply:
> about not using a lock on this cached map (writing directly to the map contents to update state from unknown to ready, and other changes, without doing some sort of lock first via an update api?

I guess you are right, I reconstructed the code of PodSandbox about the status and add a commit in this PR, please take a look.

Reply:
please take a look again. @mikebrow